### PR TITLE
feature(xjy): Supports options for selecting encoder/decoder

### PR DESCRIPTION
**Pull request by xiongjyu**
It supports using option=["qwen", "legacy"] to select qwen as the encoder/decoder or the original bge-encoder/t5-decoder

### CODE REVIEW

File: lzero/mcts/buffer/game_buffer.py [hunk 1]
```diff
@@ -156,7 +156,7 @@ def _sample_orig_data(self, batch_size: int) -> Tuple:
                 # For some environments (e.g., Jericho), the action space size may be different.
                 # To ensure we can always unroll `num_unroll_steps` steps starting from the sampled position (without exceeding segment length),
                 # we avoid sampling from the last `num_unroll_steps` steps of the game segment. 
-                if pos_in_game_segment >= self._cfg.game_segment_length - self._cfg.num_unroll_steps:
+                if pos_in_game_segment >= self._cfg.game_segment_length - self._cfg.num_unroll_steps - self._cfg.td_steps:
                     pos_in_game_segment = np.random.choice(self._cfg.game_segment_length - self._cfg.num_unroll_steps, 1).item()
```
Comment (puyuan1996) [inline lzero/mcts/buffer/game_buffer.py]: 这里的改动是否能修复bug3呢

File: lzero/model/common.py [hunk 2]
```diff
@@ -21,6 +21,8 @@
 from ditk import logging
 from ding.utils import set_pkg_seed, get_rank, get_world_size
 import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
```
Comment (puyuan1996) [inline lzero/model/common.py:25]: 按照规范，修复import的顺序

File: lzero/model/common.py [hunk 3]
```diff
@@ -361,6 +363,115 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        super().__init__()
+
+        logging.info(f"Loading Qwen model from: {model_path}")
+        
+        local_rank = get_rank()
+        if local_rank == 0:
+            self.pretrained_model = AutoModelForCausalLM.from_pretrained(
+                model_path,
+                torch_dtype="auto", 
+                device_map={"": local_rank},
```
Comment (puyuan1996) [inline lzero/model/common.py:382]: 是否应该写成cuda:{local_rank}呢？

File: lzero/model/common.py [hunk 4]
```diff
@@ -361,6 +363,115 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            Encode the input tensor `x` to a latent state.
+        Arguments:
+            - x (:obj:`torch.Tensor`): Input tensor of shape (B, C_in, W, H).
+        Returns:
+            - latent_state (:obj:`torch.Tensor`): Encoded latent state of shape (B, embedding_dim).
+        """
+        pad_id = self.tokenizer.pad_token_id
+        attention_mask = (x != pad_id).long().to(x.device)
+        context = {'input_ids': x.long(), 'attention_mask': attention_mask}
+        no_grad = True
```
Comment (puyuan1996) [inline lzero/model/common.py]: 应该由传入参数控制？

File: lzero/model/common.py [hunk 5]
```diff
@@ -361,6 +363,115 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            return nn.LayerNorm(embedding_size)
+        else:
+            raise NotImplementedError(f"Normalization type '{norm_option}' is not implemented.")
+
+    def encode(self, x: torch.Tensor, no_grad: bool = True) -> torch.Tensor:
+        """
+        Overview:
+            Encode the input tensor `x` to a latent state.
+        Arguments:
+            - x (:obj:`torch.Tensor`): Input tensor of shape (B, C_in, W, H).
```
Comment (puyuan1996) [inline lzero/model/common.py]: x的shape实际是(B, L)？

File: lzero/model/common.py [hunk 6]
```diff
@@ -361,6 +363,115 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+    def decode(self, embeddings: torch.Tensor, max_length: int = 512) -> str:
+        """
+        Decodes embeddings into text via the decoder network.
+        """
+        embeddings_detached = embeddings.detach()
+        self.pretrained_model.eval()
+        
+        # Directly generate using provided embeddings
+        with torch.no_grad():
+            param = next(self.pretrained_model.parameters())
```
Comment (puyuan1996) [inline lzero/model/common.py:463]: 为什么这里要使用 param = next(self.pretrained_model.parameters()) 来提取device呢

File: lzero/model/common.py [hunk 7]
```diff
@@ -361,6 +363,115 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            embeddings = embeddings_detached.to(device=param.device, dtype=param.dtype)
+            gen_ids = self.pretrained_model.generate(
```
Comment (puyuan1996) [inline lzero/model/common.py:465]: 目前这种方式是能正确运行，且训练后decode出的文本也是bleu很高的吗

File: lzero/model/unizero_world_models/tokenizer.py [hunk 8]
```diff
@@ -39,7 +39,7 @@ class Tokenizer(nn.Module):
         Can operate on visual or textual data, supporting optional LPIPS perceptual loss.
         It optionally includes a linear projection layer and can be paired with a decoder tokenizer.
     """
-    def __init__(self, encoder=None, decoder_network=None, decoder_network_tokenizer=None, with_lpips: bool = False, projection: list = None) -> None:
+    def __init__(self, encoder=None, decoder_network=None, decoder_network_tokenizer=None, with_lpips: bool = False, projection: list = None, encoder_option='legacy') -> None:
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/tokenizer.py:42]: encoder_option 改成 encoder_decoder_type 吧

File: lzero/model/unizero_world_models/tokenizer.py [hunk 9]
```diff
@@ -49,6 +49,7 @@ def __init__(self, encoder=None, decoder_network=None, decoder_network_tokenizer
             with_lpips (bool, optional): If True, enable perceptual loss computation via LPIPS. Defaults to False.
             projection (list[int], optional): If provided, defines a linear projection layer from projection[0] → projection[1]. 
                                               If None, an identity layer is used.
+            encoder_option (str, optional): Option to specify the encoder type, e.g., 'legacy' for T5 decoder or 'qwen' for Qwen decoder. Defaults to 'legacy'.
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/tokenizer.py:52]: 注释需要更新一下，可以改成 encoder_decoder_type

File: lzero/model/unizero_model.py [hunk 10]
```diff
@@ -216,6 +232,9 @@ def initial_inference(self, obs_batch: torch.Tensor, action_batch: Optional[torc
         reward = logits_rewards
         policy_logits = logits_policy.squeeze(1)
         value = logits_value.squeeze(1)
+        if torch.isnan(value).any() or torch.isnan(latent_state).any() or torch.isnan(policy_logits).any():
```
Comment (puyuan1996) [inline lzero/model/unizero_model.py]: 改为一段间隔再检查，减少计算开销

File: lzero/model/unizero_world_models/tokenizer.py [hunk 11]
```diff
@@ -146,34 +134,69 @@ def decode_to_reconstruction_outputs(self, embeddings: torch.Tensor, target_ids:
...
+            prompt_attention_mask = torch.ones(hidden.size(0), 1, device=param.device, dtype=torch.long)
+            target_attention_mask = (target_ids != self.decoder_network.tokenizer.pad_token_id).to(device=param.device, dtype=torch.long)
+            attention_mask = torch.cat([prompt_attention_mask, target_attention_mask], dim=1)
+
+            prompt_labels = torch.full((hidden.size(0), 1), -100, device=param.device, dtype=torch.long)
+
+            labels = target_ids.clone().to(param.device)
+            labels[labels == self.decoder_network.tokenizer.pad_token_id] = -100
+
+            final_labels = torch.cat([prompt_labels, labels], dim=1)
```
Comment (puyuan1996) [inline lzero/model/unizero_world_models/tokenizer.py:196]: 这里是由于qwen是decoder_only模型，所以需要手动给出 prompt + target 的 label 吗

File: lzero/policy/unizero.py [hunk 12]
```diff
@@ -861,10 +868,11 @@ def _reset_collect(self, env_id: int = None, current_steps: int = None, reset_in
             - reset_init_data (:obj:`bool`, optional): Whether to reset the initial data. If True, the initial data will be reset.
         """
         if reset_init_data:
-            self.last_batch_obs = initialize_zeros_batch(
+            self.last_batch_obs = initialize_pad_batch(
```
Comment (puyuan1996) [inline lzero/policy/unizero.py:875]: 需要修改为兼容之前非文本环境的情况

File: zoo/jericho/configs/jericho_unizero_ddp_config.py [hunk 13]
```diff
@@ -105,11 +110,12 @@ def main(env_id: str = 'detective.z5', seed: int = 0, max_env_step: int = int(1e
                     ),
                 ),
             ),
-            accumulation_steps=4,  # TODO: Accumulated gradient steps (currently default)
+            accumulation_steps=1,  # TODO: Accumulated gradient steps (currently default)
```
Comment (puyuan1996) [inline zoo/jericho/configs/jericho_unizero_ddp_config.py]: 因为accumulation_steps也会影响upc中的值，改成在最前面手动设定，然后分别赋值到accumulation_steps和update_per_collect里吧

File: zoo/jericho/envs/test_qwen.py [hunk 14]
```diff
@@ -0,0 +1,233 @@
+import logging
```
Comment (puyuan1996) [inline zoo/jericho/envs/test_qwen.py]: 这里和https://github.com/opendilab/LightZero/pull/385/files中的文件完全一样吗？如果是的话，可以都在这个PR中开发，哪个PR可以关了哈。以及file name改成 test_qwen_prior.py 吧

File: lzero/model/common.py (hunk 3 above)
Comment (xiongjyu) [reply lzero/model/common.py:382]: image

File: lzero/model/unizero_world_models/tokenizer.py (hunk 11 above)
Comment (xiongjyu) [reply lzero/model/unizero_world_models/tokenizer.py:196]: 对

File: lzero/policy/unizero.py (hunk 12 above)
Comment (xiongjyu) [reply lzero/policy/unizero.py:875]: 现在这样已经是兼容的了，因为非文本情况下没有tokenizer，所以self.pad_id=0
image

File: lzero/entry/utils.py [hunk 15]
```diff
@@ -111,6 +111,28 @@ def initialize_zeros_batch(observation_shape: Union[int, List[int], Tuple[int]],
 
     return torch.zeros(shape).to(device)
 
+def initialize_pad_batch(observation_shape: Union[int, List[int], Tuple[int]], batch_size: int, device: str, pad_token_id: int) -> torch.Tensor:
+    """
+    Overview:
+        Initialize a tensor filled with `pad_token_id` for batch observations. 
+        This is typically used to initialize input_ids with padding tokens.
```
Comment (puyuan1996) [inline lzero/entry/utils.py]: 注释需要更新一下，说一下是兼容文本和非文本的情况

File: lzero/policy/unizero.py [hunk 16]
```diff
@@ -348,6 +348,9 @@ def _init_learn(self) -> None:
         self.l2_norm_after = 0.
         self.grad_norm_before = 0.
         self.grad_norm_after = 0.
+
```
Comment (puyuan1996) [inline lzero/policy/unizero.py:355]: 在unizero的默认config中增加decode_loss_mode 而且默认值为'None'，注明不同option的含义

File: zoo/jericho/configs/jericho_unizero_ddp_config.py [hunk 17]
```diff
@@ -16,10 +16,11 @@ def main(env_id: str = 'detective.z5', seed: int = 0, max_env_step: int = int(1e
     Returns:
         None
     """
-    gpu_num = 2
+    gpu_num = 4
     collector_env_num: int = 4       # Number of collector environments
     n_episode = int(collector_env_num*gpu_num)
-    batch_size = int(8*gpu_num)
+    batch_size = int(1*gpu_num)
```
Comment (puyuan1996) [inline zoo/jericho/configs/jericho_unizero_ddp_config.py:22]: jericho目前的实验都是用的这个config跑的吗？segment config那个有测试过收敛性吗

File: zoo/jericho/envs/test_qwen_prior.py [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+            collector_env_num=1,
+            evaluator_env_num=1,
+            save_replay=False,
+            save_replay_path=None,
+            env_type=env_type,
+            collect_policy_mode='expert'    # random, human, expert
+        )
+    )
+
+    if env_cfg.add_location_and_inventory:
```
Comment (puyuan1996) [inline zoo/jericho/envs/test_qwen_prior.py]: 这个文件可以先删除了 后面结合https://github.com/opendilab/LightZero/pull/397的test_qwen_v9.py合并一下，对qwen-prior这个新提一个PR

File: zoo/jericho/configs/jericho_unizero_ddp_config.py (hunk 17 above)
Comment (xiongjyu) [reply zoo/jericho/configs/jericho_unizero_ddp_config.py:22]: 单任务的没有试过，但是之前多任务的跑segment不如当前这种

File: zoo/jericho/envs/test_qwen_prior.py (hunk 18 above)
Comment (xiongjyu) [reply zoo/jericho/envs/test_qwen_prior.py]: 好，那我当前就不用这个文件了，等之后这个397的pr搞好了直接merge进来
